File: ansible_collections/junioryono/billet/plugins/module_utils/endpoint.py
# Copyright (c) 2026 junioryono
# Apache-2.0

"""The one representation of a control-plane endpoint, in Python.

`internal/endpoint` is billet's Go representation: a scheme decided by the
node's TLS state, a typed host (a literal address canonicalised, or a DNS
spelling under a pinned ASCII grammar with its rootedness kept), a numeric
port, one canonical text, and equality by value. This module is the same
representation for the collection's Python, and BOTH are held to one vector
table, `tests/fixtures/endpoint-vectors.json`, row by row, never to each
other's defaults.

WHY NOT `urlsplit` ALONE. Python's `urlsplit` decides less than Go's `url.Parse`
and decides some of it differently (measured on Python 3.14.6 against Go 1.26,
2026-09-09): it strips a tab from the text where Go refuses any control
character; it keeps every `%` escape undecoded and unvalidated where Go
validates the escapes of the userinfo, the host, the path and the fragment
(never the query) and decodes the host's; it cannot tell an empty query or
fragment (`?`, `#`) from none; it answers no port for `host:` and for a
portless host alike; and `ipaddress` accepts a scoped address where Go refuses
a zone. So this module GATHERS the raw spelling's facts first, without
refusing, and then refuses in Go's order: syntax, scheme, host, route, port,
then the typed host (zone, grammar). A row the two languages could read apart
is a row of the table.
"""

from __future__ import absolute_import, division, print_function

__metaclass__ = type

import ipaddress
import re
from urllib.parse import urlsplit

KIND_LITERAL = "literal"
KIND_DNS = "dns"

_SYNTAX = "syntax"
_SCHEME = "scheme"
_HOST = "host"
_ROUTE = "route"
_ZONE = "zone"
_PORT = "port"
_GRAMMAR = "grammar"
_NOT_CANONICAL = "not_canonical"

KINDS = (_SYNTAX, _SCHEME, _HOST, _ROUTE, _ZONE, _PORT, _GRAMMAR, _NOT_CANONICAL)


class EndpointError(ValueError):
    """A spelling refused, with the one kind it is refused for."""

    def __init__(self, kind, message):
        if kind not in KINDS:
            raise ValueError("endpoint: unknown refusal kind %r" % (kind,))
        super(EndpointError, self).__init__("endpoint: %s: %s" % (kind, message))
        self.kind = kind
        self.message = message


class Endpoint(object):
    """A scheme, a typed host and a port; equal by value."""

    __slots__ = ("scheme", "kind", "host", "rooted", "port")

    def __init__(self, scheme, kind, host, rooted, port):
        self.scheme = scheme
        self.kind = kind
        self.host = host
        self.rooted = rooted
        self.port = port

    def canonical(self):
        """The canonical spelling: the scheme, the host (a literal in RFC 5952
        text, an IPv6 one bracketed; a name lower-cased with its trailing dot
        when rooted) and the port, always written."""
        if self.kind == KIND_LITERAL and ":" in self.host:
            host = "[" + self.host + "]"
        elif self.kind == KIND_DNS and self.rooted:
            host = self.host + "."
        else:
            host = self.host
        return "%s://%s:%d" % (self.scheme, host, self.port)

    def __eq__(self, other):
        if not isinstance(other, Endpoint):
            return NotImplemented
        return (self.scheme == other.scheme and self.kind == other.kind and self.host == other.host
                and self.rooted == other.rooted and self.port == other.port)

    def __ne__(self, other):
        eq = self.__eq__(other)
        if eq is NotImplemented:
            return eq
        return not eq

    def __hash__(self):
        return hash((self.scheme, self.kind, self.host, self.rooted, self.port))

    def __repr__(self):
        return "Endpoint(%r)" % (self.canonical(),)


_HEX = "0123456789abcdefABCDEF"


def _escapes(segment, where):
    """Every `%` in the segment starts a two-hex-digit escape, as Go's URL
    parser requires of the userinfo, the path and the fragment; the host's
    stricter rule is in `_host_escapes`."""
    i = 0
    while i < len(segment):
        if segment[i] == "%":
            if len(segment) - i < 3 or segment[i + 1] not in _HEX or segment[i + 2] not in _HEX:
                raise EndpointError(_SYNTAX, "not a URL: invalid URL escape %r in the %s" % (segment[i:i + 3], where))
            i += 3
        else:
            i += 1


def _host_escapes(host):
    """Go's host rule: an escape is admitted only for `%25` and for a byte
    that is not ASCII (the way a non-ASCII name is written percent-encoded),
    and every admitted escape is DECODED before the host is typed."""
    out = bytearray()
    i = 0
    while i < len(host):
        if host[i] == "%":
            if len(host) - i < 3 or host[i + 1] not in _HEX or host[i + 2] not in _HEX:
                raise EndpointError(_SYNTAX, "not a URL: invalid URL escape %r in the host" % (host[i:i + 3],))
            value = int(host[i + 1:i + 3], 16)
            if value != 0x25 and value < 0x80:
                raise EndpointError(_SYNTAX, "not a URL: invalid URL escape %r in the host" % (host[i:i + 3],))
            out.append(value)
            i += 3
        else:
            out.extend(host[i].encode("utf-8"))
            i += 1
    return out.decode("utf-8", "surrogateescape")


def parse(text, tls):
    """Read a spelling under the node's TLS state: a bare `host:port`, a bare
    host (the scheme's default port), or a URL whose scheme must agree with
    the TLS state. Raises EndpointError with one kind."""
    if not isinstance(text, str):
        raise EndpointError(_SYNTAX, "not a URL: %r is not text" % (text,))

    scheme = "https" if tls else "http"
    raw = text if "://" in text else scheme + "://" + text

    # SYNTAX, gathered from the raw spelling before anything is refused for a
    # later reason, in the order Go's parser meets it.
    for c in raw:
        if ord(c) < 0x20 or ord(c) == 0x7F:
            raise EndpointError(_SYNTAX, "not a URL: %r carries a control character" % (text,))

    explicit, _, rest = raw.partition("://")
    if not explicit or not re.match(r"^[A-Za-z][A-Za-z0-9+.-]*$", explicit):
        raise EndpointError(_SYNTAX, "not a URL: %r has no scheme" % (text,))
    explicit = explicit.lower()

    # The authority ends at the first `/`, `?` or `#`; what follows is routing.
    end = len(rest)
    for stop in "/?#":
        at = rest.find(stop)
        if at != -1 and at < end:
            end = at
    authority, remainder = rest[:end], rest[end:]

    userinfo = None
    hostport = authority
    if "@" in authority:
        userinfo, _, hostport = authority.rpartition("@")
        _escapes(userinfo, "userinfo")

    if hostport.startswith("["):
        close = hostport.find("]")
        if close == -1:
            raise EndpointError(_SYNTAX, "not a URL: %r has an unclosed bracket" % (text,))
        host_raw, after = hostport[1:close], hostport[close + 1:]
        if after and not after.startswith(":"):
            raise EndpointError(_SYNTAX, "not a URL: %r has text after the bracketed host" % (text,))
        port_raw = after[1:] if after else None
        bracketed = True
    else:
        if hostport.count(":") > 1:
            raise EndpointError(_SYNTAX, "not a URL: %r has an invalid port after its host" % (text,))
        host_raw, colon, port_raw = hostport.partition(":")
        if not colon:
            port_raw = None
        bracketed = False

    if port_raw and not port_raw.isdigit():
        raise EndpointError(_SYNTAX, "not a URL: %r has an invalid port %r after its host" % (text, ":" + port_raw))

    if bracketed:
        # Go's parser validates a bracketed host as an IP literal, a zone
        # written `%25...` with a non-empty name; the escape is decoded here.
        host_decoded = _host_escapes(host_raw) if "%" in host_raw else host_raw
        if "%" in host_decoded:
            addr_text, _, zone = host_decoded.partition("%")
            if zone == "":
                raise EndpointError(_SYNTAX, "not a URL: %r: zone must be a non-empty string" % (text,))
        else:
            addr_text = host_decoded
        try:
            literal = ipaddress.ip_address(addr_text)
        except ValueError:
            raise EndpointError(_SYNTAX, "not a URL: %r has an invalid host %r" % (text, host_raw))
        if not isinstance(literal, ipaddress.IPv6Address):
            raise EndpointError(_SYNTAX, "not a URL: %r has an invalid IP-literal" % (text,))
    else:
        host_decoded = _host_escapes(host_raw) if "%" in host_raw else host_raw

    # Try the whole spelling through urlsplit too: what it cannot parse is a
    # syntax refusal here as it is in Go.
    try:
        split = urlsplit(raw)
        _ = split.hostname
    except ValueError as exc:
        raise EndpointError(_SYNTAX, "not a URL: %r: %s" % (text, exc))

    if remainder.startswith("/"):
        path, _, tail = remainder.partition("?")
        path, _, _ = path.partition("#")
        _escapes(path, "path")
    fragment_at = remainder.find("#")
    if fragment_at != -1:
        _escapes(remainder[fragment_at + 1:], "fragment")

    # SCHEME.
    if explicit not in ("http", "https"):
        raise EndpointError(_SCHEME, "%r must be http or https" % (text,))
    if explicit != scheme:
        raise EndpointError(_SCHEME, "%r is %s, which contradicts a node that %s; drop the scheme or write %s"
                            % (text, explicit, "has a certificate to present" if tls else "has no certificate to present", scheme))

    # HOST, the authority as a whole.
    if authority == "":
        raise EndpointError(_HOST, "%r names no host" % (text,))

    # ROUTE.
    if userinfo is not None:
        raise EndpointError(_ROUTE, "%r carries userinfo" % (text,))
    if remainder.startswith("/"):
        raise EndpointError(_ROUTE, "%r carries a path" % (text,))
    if remainder.startswith("?"):
        raise EndpointError(_ROUTE, "%r carries a query" % (text,))
    if remainder.startswith("#"):
        raise EndpointError(_ROUTE, "%r carries a fragment" % (text,))

    # PORT.
    if port_raw == "":
        raise EndpointError(_PORT, "%r has an empty port" % (text,))
    if host_decoded == "":
        raise EndpointError(_HOST, "%r names no host" % (text,))
    if port_raw is None:
        port = 443 if scheme == "https" else 80
    else:
        port = int(port_raw, 10)
        if port < 1 or port > 65535:
            raise EndpointError(_PORT, "%r: port %d is outside 1 to 65535" % (text, port))

    kind, host, rooted = _parse_host(host_decoded, text)
    return Endpoint(scheme, kind, host, rooted, port)


_LABEL = re.compile(r"^[a-z0-9-]*$")


def _parse_host(host, text):
    """Type a host: a literal when `ipaddress` accepts it (a zone refused
    before any unmapping, never stripped; a mapped address unmapped), else a
    DNS spelling under the grammar."""
    try:
        literal = ipaddress.ip_address(host)
    except ValueError:
        literal = None

    if literal is not None:
        zone = getattr(literal, "scope_id", None)
        if zone:
            raise EndpointError(_ZONE, "%r carries the zone %r, which names an interface of one host and is not an endpoint (%r)"
                                % (host, zone, text))
        mapped = getattr(literal, "ipv4_mapped", None)
        if mapped is not None:
            literal = mapped
        if isinstance(literal, ipaddress.IPv6Address):
            return KIND_LITERAL, literal.compressed, False
        return KIND_LITERAL, str(literal), False

    name = host
    rooted = False
    if name.endswith("."):
        rooted = True
        name = name[:-1]

    for c in name:
        if ord(c) >= 0x80:
            raise EndpointError(_GRAMMAR, "name: %r is not ASCII; an internationalised name is admitted only as its xn-- form (%r)"
                                % (host, text))

    name = name.lower()

    for label in name.split("."):
        if label == "":
            raise EndpointError(_GRAMMAR, "name: %r has an empty label (%r)" % (host, text))
        if label.startswith("-"):
            raise EndpointError(_GRAMMAR, "name: the label %r of %r begins with a hyphen (%r)" % (label, host, text))
        if label.endswith("-"):
            raise EndpointError(_GRAMMAR, "name: the label %r of %r ends with a hyphen (%r)" % (label, host, text))
        for c in label:
            if c == "_":
                raise EndpointError(_GRAMMAR, "name: %r carries an underscore, which is not a hostname (%r)" % (host, text))
            if not _LABEL.match(c):
                raise EndpointError(_GRAMMAR, "name: %r carries %r, which is not a hostname character (%r)" % (host, c, text))

    return KIND_DNS, name, rooted


def parse_canonical(text):
    """Read a text that must be an endpoint's own canonical spelling, under
    the scheme it carries."""
    if not isinstance(text, str):
        raise EndpointError(_NOT_CANONICAL, "%r is not text" % (text,))
    tls = text.startswith("https://")
    if not tls and not text.startswith("http://"):
        raise EndpointError(_NOT_CANONICAL, "%r carries no scheme" % (text,))
    endpoint = parse(text, tls)
    if endpoint.canonical() != text:
        raise EndpointError(_NOT_CANONICAL, "%r spells %r" % (text, endpoint.canonical()))
    return endpoint

File: internal/endpoint/endpoint.go
// Package endpoint is the one representation of a control-plane endpoint: a
// scheme, a typed host and a numeric port, parsed from the spellings a
// configuration or a record may use, printed in one canonical text, and
// compared by value. Two things read it: the node client, whose request base
// is this representation's text, and the node's registration record, which the
// inspector compares with the installed configuration. The rules here are the
// contract both billet's Go and its Python are held to, through the vector
// table in endpoint_test.go, never through each other's defaults.
//
// A HOST IS A LITERAL ADDRESS OR A DNS SPELLING, and the two never compare
// equal: `localhost` is a name whose resolution is nobody's business here, and
// no name is ever resolved. A literal is canonicalised (RFC 5952 text; an
// IPv4-mapped address is its IPv4; a zone is refused, never stripped). A name
// is held to a pinned ASCII grammar and lower-cased, and keeps its ROOTEDNESS:
// `control.example.` is a different request base from `control.example`,
// because a resolver searches the unrooted one and takes the rooted one as
// absolute. The scheme is decided by the node's TLS state alone, never by the
// address or the port. A path, userinfo, a query or a fragment refuses: this is
// an endpoint, and a routing component is never silently discarded.
package endpoint

import (
	"errors"
	"fmt"
	"net"
	"net/netip"
	"net/url"
	"strconv"
	"strings"
)

// The reasons a spelling is refused, each its own error so a caller and a test
// can tell an empty host from a port out of range from a zone.
var (
	// ErrSyntax is a spelling the URL parser refuses (an unbracketed IPv6
	// address, a raw `%` in a host, a non-digit port).
	ErrSyntax = errors.New("endpoint: not a URL")
	// ErrScheme is an explicit scheme that is neither http nor https, or one
	// that contradicts the node's TLS state.
	ErrScheme = errors.New("endpoint: scheme")
	// ErrHost is a spelling with no host.
	ErrHost = errors.New("endpoint: no host")
	// ErrRoute is a routing component beside the endpoint: a path, userinfo, a
	// query or a fragment, empty ones included.
	ErrRoute = errors.New("endpoint: a routing component")
	// ErrZone is a literal address with a zone.
	ErrZone = errors.New("endpoint: a zone")
	// ErrPort is a port that is empty or outside 1 to 65535.
	ErrPort = errors.New("endpoint: port")
	// ErrGrammar is a DNS spelling outside the pinned grammar.
	ErrGrammar = errors.New("endpoint: name")
	// ErrNotCanonical is a text ParseCanonical was given that is not the
	// canonical spelling of the endpoint it names.
	ErrNotCanonical = errors.New("endpoint: not the canonical spelling")
)

// HostKind types a host.
type HostKind int

const (
	// HostLiteral is an IP address.
	HostLiteral HostKind = iota + 1
	// HostDNS is a name, never resolved here.
	HostDNS
)

// String names a kind for a message.
func (k HostKind) String() string {
	switch k {
	case HostLiteral:
		return "literal"
	case HostDNS:
		return "dns"
	}

	return fmt.Sprintf("HostKind(%d)", int(k))
}

// Host is the typed host of an endpoint: Addr for a literal (canonical,
// unmapped, no zone), Name and Rooted for a DNS spelling (lower-cased, the
// trailing dot removed and remembered).
type Host struct {
	Kind   HostKind
	Addr   netip.Addr
	Name   string
	Rooted bool
}

// Endpoint is one representation of a control-plane endpoint.
type Endpoint struct {
	Scheme string
	Host   Host
	Port   uint16
}

// Parse reads a spelling under the node's TLS state: a bare `host:port`, a
// bare host (the scheme's default port), or a URL whose scheme must agree with
// the TLS state.
func Parse(input string, tls bool) (Endpoint, error) {
	scheme := "http"
	if tls {
		scheme = "https"
	}

	raw := input
	if !strings.Contains(raw, "://") {
		raw = scheme + "://" + raw
	}

	u, err := url.Parse(raw)
	if err != nil {
		return Endpoint{}, fmt.Errorf("%w: %q: %w", ErrSyntax, input, err)
	}

	switch u.Scheme {
	case "http", "https":
	default:
		return Endpoint{}, fmt.Errorf("%w: %q must be http or https", ErrScheme, input)
	}

	if u.Scheme != scheme {
		return Endpoint{}, fmt.Errorf("%w: %q is %s, which contradicts a node that %s; drop the scheme or write %s",
			ErrScheme, input, u.Scheme, tlsState(tls), scheme)
	}

	if u.Host == "" {
		return Endpoint{}, fmt.Errorf("%w: %q names no host", ErrHost, input)
	}

	switch {
	case u.User != nil:
		return Endpoint{}, fmt.Errorf("%w: %q carries userinfo", ErrRoute, input)
	case u.Path != "" || u.RawPath != "":
		return Endpoint{}, fmt.Errorf("%w: %q carries a path", ErrRoute, input)
	case u.RawQuery != "" || u.ForceQuery:
		return Endpoint{}, fmt.Errorf("%w: %q carries a query", ErrRoute, input)
	case u.Fragment != "" || strings.HasSuffix(raw, "#"):
		return Endpoint{}, fmt.Errorf("%w: %q carries a fragment", ErrRoute, input)
	}

	// A COLON WITH NOTHING AFTER IT IS AN EMPTY PORT, which the splitter
	// reads as no port at all; no port at all is the scheme's default,
	// because that is what the client dials there.
	if strings.HasSuffix(u.Host, ":") {
		return Endpoint{}, fmt.Errorf("%w: %q has an empty port", ErrPort, input)
	}

	hostPart, portPart, err := net.SplitHostPort(u.Host)
	if err != nil {
		hostPart = strings.TrimSuffix(strings.TrimPrefix(u.Host, "["), "]")
		portPart = ""
	}

	if hostPart == "" {
		return Endpoint{}, fmt.Errorf("%w: %q names no host", ErrHost, input)
	}

	port, err := parsePort(portPart, scheme)
	if err != nil {
		return Endpoint{}, fmt.Errorf("%w: %q: %w", ErrPort, input, err)
	}

	host, err := parseHost(hostPart)
	if err != nil {
		return Endpoint{}, fmt.Errorf("%w (%q)", err, input)
	}

	return Endpoint{Scheme: scheme, Host: host, Port: port}, nil
}

// ParseCanonical reads a text that must be an endpoint's own canonical
// spelling, under the scheme it carries: what the registration record holds,
// and what a comparison with a configuration is made over.
func ParseCanonical(text string) (Endpoint, error) {
	tls := strings.HasPrefix(text, "https://")
	if !tls && !strings.HasPrefix(text, "http://") {
		return Endpoint{}, fmt.Errorf("%w: %q carries no scheme", ErrNotCanonical, text)
	}

	e, err := Parse(text, tls)
	if err != nil {
		return Endpoint{}, err
	}

	if e.String() != text {
		return Endpoint{}, fmt.Errorf("%w: %q spells %q", ErrNotCanonical, text, e.String())
	}

	return e, nil
}

// String is the canonical spelling: the scheme, the host (a literal in RFC
// 5952 text, an IPv6 one bracketed; a name lower-cased with its trailing dot
// when rooted) and the port, always written.
func (e Endpoint) String() string {
	var host string

	switch e.Host.Kind {
	case HostLiteral:
		if e.Host.Addr.Is6() {
			host = "[" + e.Host.Addr.String() + "]"
		} else {
			host = e.Host.Addr.String()
		}
	case HostDNS:
		host = e.Host.Name
		if e.Host.Rooted {
			host += "."
		}
	}

	return e.Scheme + "://" + host + ":" + strconv.Itoa(int(e.Port))
}

// Equal is equality of scheme, typed host and port.
func (e Endpoint) Equal(o Endpoint) bool {
	if e.Scheme != o.Scheme || e.Port != o.Port || e.Host.Kind != o.Host.Kind {
		return false
	}

	switch e.Host.Kind {
	case HostLiteral:
		return e.Host.Addr == o.Host.Addr
	case HostDNS:
		return e.Host.Name == o.Host.Name && e.Host.Rooted == o.Host.Rooted
	}

	return false
}

func tlsState(tls bool) string {
	if tls {
		return "has a certificate to present"
	}

	return "has no certificate to present"
}

// parsePort is decimal digits between 1 and 65535, or the scheme's default
// when there are none. The URL parser has already refused a non-digit.
func parsePort(text, scheme string) (uint16, error) {
	if text == "" {
		if scheme == "https" {
			return 443, nil
		}

		return 80, nil
	}

	n, err := strconv.ParseUint(text, 10, 64)
	if err != nil {
		return 0, fmt.Errorf("port %q is not a number", text)
	}

	if n < 1 || n > 65535 {
		return 0, fmt.Errorf("port %d is outside 1 to 65535", n)
	}

	return uint16(n), nil
}

// parseHost types a host: a literal when netip accepts it (a zone refused
// before any unmapping, never stripped; a mapped address unmapped), else a DNS
// spelling under the grammar.
func parseHost(text string) (Host, error) {
	if addr, err := netip.ParseAddr(text); err == nil {
		if addr.Zone() != "" {
			return Host{}, fmt.Errorf("%w: %q carries the zone %q, which names an interface of one host and is not an endpoint",
				ErrZone, text, addr.Zone())
		}

		return Host{Kind: HostLiteral, Addr: addr.Unmap()}, nil
	}

	name := text
	rooted := false

	if strings.HasSuffix(name, ".") {
		rooted = true
		name = strings.TrimSuffix(name, ".")
	}

	for i := range len(name) {
		if name[i] >= 0x80 {
			return Host{}, fmt.Errorf("%w: %q is not ASCII; an internationalised name is admitted only as its xn-- form",
				ErrGrammar, text)
		}
	}

	name = strings.ToLower(name)

	for label := range strings.SplitSeq(name, ".") {
		if label == "" {
			return Host{}, fmt.Errorf("%w: %q has an empty label", ErrGrammar, text)
		}

		if strings.HasPrefix(label, "-") {
			return Host{}, fmt.Errorf("%w: the label %q of %q begins with a hyphen", ErrGrammar, label, text)
		}

		if strings.HasSuffix(label, "-") {
			return Host{}, fmt.Errorf("%w: the label %q of %q ends with a hyphen", ErrGrammar, label, text)
		}

		for i := range len(label) {
			c := label[i]

			switch {
			case c >= 'a' && c <= 'z', c >= '0' && c <= '9', c == '-':
			case c == '_':
				return Host{}, fmt.Errorf("%w: %q carries an underscore, which is not a hostname", ErrGrammar, text)
			default:
				return Host{}, fmt.Errorf("%w: %q carries %q, which is not a hostname character", ErrGrammar, text, c)
			}
		}
	}

	return Host{Kind: HostDNS, Name: name, Rooted: rooted}, nil
}
